### ISSUE TITLE
Wrong indentation on jsdoc block

### ISSUE THREAD

**Issue by kristijanhusak** [nvim-treesitter/nvim-treesitter#1167]
**Describe the bug**

Indentation in jsdoc blocks is completely off when treesitter indents are enabled.

**To Reproduce**

1. Open new javascript file
2. Add this content
```javascript
class Test {
  constructor(one, two) {

  }
}
```
3. Go above the constructor, start typing jsdoc block (`/**`)
4. do `<CR>`, see that indentation is on wrong place. Example:
```javascript
class Test {
  /**
* @param {String} one
*/
  constructor(one, two) {

  }
}
```
5. Try to reindent with `=`, it messes up even more. Example:
```javascript
class Test {
/**
* @param {String} one
  */
  constructor(one, two) {

  }
}
```

**Expected behavior**

I expect docblock to be properly formatted on `<CR>` and `=`. Basically this:
```javascript
class Test {
  /**
   * @param {String} one
   */
  constructor(one, two) {

  }
}
```

**Output of `:checkhealth nvim_treesitter`**

<details>
<code>

health#nvim_treesitter#check
========================================================================
## Installation
  - OK: `tree-sitter` found  0.19.3 (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v12.2.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl" }
  - OK: Neovim was compiled with tree-sitter runtime ABI version 13 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I
  - php            ✓ ✓ ✓ ✓ 
  - python         ✓ ✓ ✓ ✓ 
  - html           ✓ ✓ ✓ ✓ 
  - tsx            ✓ ✓ ✓ ✓ 
  - jsdoc          ✓ . . . 
  - c              ✓ ✓ ✓ ✓ 
  - lua            ✓ ✓ ✓ ✓ 
  - typescript     ✓ ✓ ✓ ✓ 
  - javascript     ✓ ✓ ✓ ✓ 
  - go             ✓ ✓ ✓ ✓ 

 Legend: H[ighlight], L[ocals], F[olds], I[ndents]
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}

</code>
</details>

**Output of `nvim --version`**

<details>
<code>

NVIM v0.5.0-dev+1200-g3d25a72a6
Build type: Release
LuaJIT 2.1.0-beta3

</code>
</details>

**Additional context**

Indents without treesitter indents is also not working perfectly, if that helps.
Also, I assume this is not working because for `jsdoc` parser we have only highlights, no indents.
I'd help on this, but all these `scm` files and adding custom queries are super confusing to me.
Having a short explanation in wiki how to contribute and how to add some of these functionalities would be super helpful.

**Comment by kyazdani42** [comment]
injected parsers are not yet properly implemented, i will push a fix very soon

**Comment by kyazdani42** [comment]
i'd need help for @steelsojka. Is there any way to get the node containing the tree/parser of an injected region ?

**Comment by theHamsta** [comment]
> i'd need help for @steelsojka. Is there any way to get the node containing the tree/parser of an injected region ?

Would be good to have a proper solution for this. The best hack that I had for this was to ask all other trees about its :named_descendant_for_range and take the smallest one (meassured by node_length).

**Comment by theHamsta** [comment]
@stsewd has the same problem for incremental selection. I guess @steelsojka knows the best solution for it.

**Comment by steelsojka** [comment]
@theHamsta You can get the language tree for a range by calling

```
local parser = vim.treesitter.get_parser(bufnr)
local cursor = vim.api.nvim_win_get_cursor(0)

local parser_at_range = parser:language_for_range({cursor[1] - 1, cursor[2], cursor[1] - 1, cursor[2]})
```

This would give you the parser at whatever range you need. Language trees don't have references to their parents, but they probably should. You could get the parent by iterating from the root parser and keep track of the parent while iterating. You can use `language_tree:for_each_child(function(lang_tree) do_stuff() end)`

Now a language tree can have multiple root nodes for you need to iterate over all roots in the tree using `language_tree:trees()` to get a list of all the root node.

**Comment by theHamsta** [comment]
`local parser_at_range = parser:language_for_range({cursor[1] - 1, cursor[2], cursor[1] - 1, cursor[2]})`

but this will give you the parser at this range. But we want its parent. I also thought about saving a reference to the parent when I was looking for a solution.

**Comment by steelsojka** [comment]
Yeah I think the simplest thing is to add the parent as an argument to the language tree options. That way we can have a parent method that returns the parent. I wouldn't ever change so it should be fine. I can submit a PR tomorrow unless someone beats me to it.

**Comment by arowe92** [comment]
Just noting I get the same behavior for block comments in c++. 
I had to disable indent for treesitter, indentation is correct without it.

**Comment by kevinkleine** [comment]
This works for me after installing jsdoc language.

**Comment by kristijanhusak** [comment]
@kevinkleine are you sure you are using treesitter indent? What's the output of `echo &indentexpr`?

I'm still getting this result:
```js
class Test {
  /**
* @param {String} one
*/
  constructor(one, two) {

  }
}
```

**Comment by kristijanhusak** [comment]
I wrote a workaround in my config until this gets resolved. Here's to anyone who needs it:
```lua
function _G.javascript_indent()
  local line = vim.fn.getline(vim.v.lnum)
  local prev_line = vim.fn.getline(vim.v.lnum - 1)
  if line:match('^%s*[%*/]%s*') then
    if prev_line:match('^%s*%*%s*') then
      return vim.fn.indent(vim.v.lnum - 1)
    end
    if prev_line:match('^%s*/%*%*%s*$') then
      return vim.fn.indent(vim.v.lnum - 1) + 1
    end
  end

  return vim.fn['GetJavascriptIndent']()
end

vim.cmd[[autocmd FileType javascript setlocal indentexpr=v:lua.javascript_indent()]]
```

**Comment by kevinkleine** [comment]
@kristijanhusak 

GetJavascriptIndent()

I also have nvim-treesitter installed with the jsdoc language, perhaps that is why it is indenting properly here. (I can post my init.vim if you want)

**Comment by kristijanhusak** [comment]
@kevinkleine you are probably not using treesitter highlights, that's why it's working for you. Js indent function heavily relies on the syntax groups used on the current line to determine the indentation, which is not working when treesitter highlights are enabled (syntax is off in that case).

**Comment by danielnehrig** [comment]
is there any updates on this issue is still relevant

**Comment by DRogueRonin** [comment]
I have the exact same problem with php doc comments. The lines after the opening `/**` are not automatically indented and get pushed back when auto-formatting the whole file.

**Comment by MarkoSagadin** [comment]
Chiming in: I am also having problems with mulitline comments in C/C++. Disabling tree-sitter indenting on .c and c++ fixes the indenting problem.

It is weird, I have been using the same setup for a month now, and only noticed an issue now, I did not touch my neovim config.

**Comment by MarkoSagadin** [comment]
Aha, I understand.

**Comment by FernandoBasso** [comment]
I my case, I also get this kind of “wrong” indentation with JS and TS:

![image](https://user-images.githubusercontent.com/6560620/154658707-58d7a553-7d4c-4193-8ba5-0badee408d3e.png)

**Comment by aspeddro** [comment]
I added comment as auto in `ecma`:

```sheme
(comment) @auto
```

https://user-images.githubusercontent.com/16160544/155643011-1b1c6c9c-ca06-42d9-b62f-e6b9ae7241ba.mp4

**Comment by FernandoBasso** [comment]
@aspeddro We cannot watch the video because Github says it is corrupt.

**Comment by aspeddro** [comment]
> @aspeddro We cannot watch the video because Github says it is corrupt.
> 
> Also, can you explain your previous comment in more detail?

When updating the `comment` node in the query to `auto` the indentation worked as expected.

1. Create `indents.scm` at `~/.config/nvim/after/queries/ecma/indents.scm`
2. Add:
   ```scheme
   (comment) @auto
   ```

> By default `comment` is set to `ignore`

https://github.com/nvim-treesitter/nvim-treesitter/blob/2f704e538c2135a9a4628c19b94dec316e238197/queries/ecma/indents.scm#L48-L51

![image](https://user-images.githubusercontent.com/16160544/159998072-e8f826b4-d18f-4a8c-b46e-21aaed5f68bd.png)

**Comment by icholy** [comment]
Setting `(comment) @auto` works for indenting a doc block if it's already closed.

``` ts
/**
 * <- hitting [enter] here will work correctly
 */
```

If the block isn't closed yet, there indentation doesn't work.

``` ts
/**
 * <- hitting [enter] here will not work correctly.
```

**Comment by aspeddro** [comment]
yes, i always use closed (snippet closes automatically). If you haven't closed the node it's not a `comment` and that's why `@auto` doesn't work.

You can map this case.

```scheme
(comment) @auto

(ERROR
 ((regex_pattern) @auto
  (#lua-match? @auto "\*\*"))) @auto
```

**Comment by aspeddro** [comment]
yes, I always use closed (snippet closes automatically). If you haven't closed the node it's not a `comment`.

**Comment by icholy** [comment]
Could there be a `partial-comment` which would make my second example work?

**Comment by aspeddro** [comment]
Depends on how parser works. Currently the second case is very difficult to map, because the parser does not treat as a special node as `jsdoc` or `partial-document`

**Comment by bennypowers** [comment]
> Yeah I think the simplest thing is to add the parent as an argument to the language tree options. That way we can have a parent method that returns the parent. I wouldn't ever change so it should be fine. I can submit a PR tomorrow unless someone beats me to it.

Sidebar: This would be really useful for regexplainer. Ideally one could traverse both ways, down into a embedded language from it's host node, and up out to the host language, similar to traversing across shadow boundaries in the [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM).

**Comment by talbergs** [comment]
Not even a work-around?

**Comment by cincodenada** [comment]
Hello, I'm having this same issue, but I'm not using the jsdoc plugin at all - I see the same behavior for regular block comments with the `javascript` plugin's indent functionality. Same behavior: after the first line, asterisks are shoved over to the first column regardless of indent, both on `<cr>` and with `=`:

```
  /*
* This is not the correct indenting!
*/
```

This is especially frustrating because the indenter I was using, `vim-javascript`, depends on the syntax tree, and thus can't coexist with Treesitter's syntax, so I have to either use Treesitter for syntax _and_ indent, or neither.

**Comment by nyngwang** [comment]
@aspeddro I still got an extra indent after closing the multiline comment block, any advice?

![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/24765272/54731a0a-e06f-46fe-9c6b-397dc866994f)
